main/selectors.c

```diff
@@ -81,11 +101,48 @@ tastePerlLine (const char *line)
 const char *
 selectByPickingPerlVersion (FILE *input)
 {
-    char line[0x800];
-    while (fgets(line, sizeof(line), input)) {
-        const char *lang = tastePerlLine(line);
-        if (lang)
-            return lang;
+    /* Default to Perl 5 */
+    return selectByLines (input, tastePerlLine, TR_PERL5);
+}
+
+static const char *
+tasteObjectiveCOrMatLabLines (const char *line)
+{
+#define TR_OBJC "ObjectiveC"
+#define TR_MATLAB "MatLab"
+    if (startsWith (line, "% ")
+	|| startsWith (line, "%{"))
+	return TR_MATLAB;
+    else if (startsWith (line, "// ")
+	     || startsWith (line, "/* "))
+	return TR_OBJC;
+    else if (startsWith (line, "#include")
+	     || startsWith (line, "#import")
+	     || startsWith (line, "#define ")
+	     || startsWith (line, "#ifdef "))
+	return TR_OBJC;
+    else if (startsWith (line, "@interface ")
+	     || startsWith (line, "@implementation ")
+	     || startsWith (line, "@protocol "))
+	return TR_OBJC;
+    else if (startsWith (line, "struct ")
+	     || startsWith (line, "union ")
+	     || startsWith (line, "typedef "))
+	return TR_OBJC;
+    else {
+	if (startsWith (line, "function ")) {
+	    const char *p = line + strlen ("function ");
+	    while (isspace(*p))
+		p++;
+	    if (*p != '\0' && *p != '(')
+		return TR_MATLAB;
+	}
```

Comment:
I like this approach! :)
Out of curiosity, did you compare the run time of both cases?